README.md
# Grid Layout (CSS) for Flutter

[![Test Status](https://github.com/lyskouski/flutter_grid_layout/actions/workflows/test.yml/badge.svg)](https://github.com/lyskouski/flutter_grid_layout/actions/workflows/test.yml)
[![Build Status](https://github.com/lyskouski/flutter_grid_layout/actions/workflows/build.yml/badge.svg)](https://github.com/lyskouski/flutter_grid_layout/actions/workflows/build.yml)
<a href="https://www.buymeacoffee.com/lyskouski"><img height="20" src="https://img.buymeacoffee.com/button-api/?text=Buy me a coffee&emoji=&slug=lyskouski&button_colour=FFDD00&font_colour=000000&font_family=Cookie&outline_colour=000000&coffee_colour=ffffff" /></a>

![Grid Layout](./assets/grid-layout.svg)

Embracing a grid layout is essential for creating adaptable and user-friendly digital designs. Grids ensure visual consistency, prioritize content, and seamlessly adapt to various screen sizes, enhancing the overall user experience across devices and displays.

## Features

- Basic implementation of [CSS Grid](https://css-tricks.com/snippets/css/complete-guide-grid/).

## Getting started
https://pub.dev/packages/flutter_grid_layout

```bash
flutter pub add flutter_grid_layout
```

## Simple Sample

```dart
GridContainer(
  columns: [0.2, null, 0.2],
  rows: [0.2, null, 0.2],
  children: [
    GridItem(
      start: const Size(1, 1),
      end: const Size(2, 2),
      child: Container(color: Colors.red),
    )
  ],
),
```

![Centered Item](https://raw.githubusercontent.com/lyskouski/flutter_grid_layout/main/test/src/grid_container_test.dart.1.png)

## Multiple Items with different layers

```dart
GridContainer(
  columns: [0.2, 0.3, 0.3, 0.2],
  rows: [0.2, 0.3, 0.3, 0.2],
  children: [
    GridItem(
      start: const Size(0, 0),
      end: const Size(4, 1),
      child: Container(color: Colors.red),
    ),
    GridItem(
      start: const Size(1, 0),
      end: const Size(3, 4),
      order: 1,
      child: Container(color: Colors.blue.withOpacity(0.5)),
    ),
    GridItem(
      start: const Size(2, 3),
      end: const Size(4, 4),
      child: Container(color: Colors.green),
    ),
  ],
),
```

![Multiple Items](https://raw.githubusercontent.com/lyskouski/flutter_grid_layout/main/test/src/grid_container_test.dart.2.png)

## Reversed Multiple Items with different layers

```dart
GridContainer(
  alignment: MainAxisAlignment.end,
  columns: [0.2, 0.3, 0.3, 0.2],
  rows: [0.2, 0.3, 0.3, 0.2],
  children: [
    GridItem(
      start: const Size(0, 0),
      end: const Size(4, 1),
      child: Container(color: Colors.red),
    ),
    GridItem(
      start: const Size(1, 0),
      end: const Size(3, 4),
      order: 1,
      child: Container(color: Colors.blue.withOpacity(0.5)),
    ),
    GridItem(
      start: const Size(2, 3),
      end: const Size(4, 4),
      child: Container(color: Colors.green),
    ),
  ],
),
```

![Multiple Items - Reversed](https://raw.githubusercontent.com/lyskouski/flutter_grid_layout/main/test/src/grid_container_test.dart.3.png)

In addition to thanking, you may [treat us to :coffee:](https://www.buymeacoffee.com/lyskouski).


CHANGELOG.md
## 1.1.1

* Reverse only horizontally

## 1.1.0

* Add support to reversed order

## 1.0.4

* Hide elements with negative values

## 1.0.3

* Update README.md

## 1.0.2

* Adjust example project to one file


## 1.0.1

* Add example project

## 1.0.0

* Use safe position binding (to avoid out-of-range error)

## 0.1.0

* Reconfigure automated publishing

## 0.0.5

* Downgrade SDK to 3.1.0

## 0.0.4

* Improve automated publishing

## 0.0.3

* Automated publishing

## 0.0.2

* Automated verification

## 0.0.1

* Basic implementation of Grid Layout


test/src/grid_container_test.dart
// Copyright 2023 The terCAD team. All rights reserved.
// Use of this source code is governed by a MIT license
// that can be found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:flutter_grid_layout/src/grid_container.dart';
import 'package:flutter_grid_layout/src/grid_item.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('ForegroundChartPainter', () {
    final testCases = [
      (
        title: 'Plot an empty scope',
        alignment: MainAxisAlignment.start,
        columns: [],
        rows: [],
        children: [],
      ),
      (
        title: 'Plot centered item',
        alignment: MainAxisAlignment.start,
        columns: [0.2, null, 0.2],
        rows: [0.2, null, 0.2],
        children: [
          GridItem(
            start: const Size(1, 1),
            end: const Size(2, 2),
            child: Container(color: Colors.red),
          )
        ],
      ),
      (
        title: 'Plot multiple nested items',
        alignment: MainAxisAlignment.start,
        columns: [0.2, 0.3, 0.3, 0.2],
        rows: [0.2, 0.3, 0.3, 0.2],
        children: [
          GridItem(
            start: const Size(0, 0),
            end: const Size(4, 1),
            child: Container(color: Colors.red),
          ),
          GridItem(
            start: const Size(1, 0),
            end: const Size(3, 4),
            order: 1,
            child: Container(color: Colors.blue.withOpacity(0.5)),
          ),
          GridItem(
            start: const Size(2, 3),
            end: const Size(4, 4),
            child: Container(color: Colors.green),
          ),
        ],
      ),
      (
        title: 'Plot multiple nested items - reversed',
        alignment: MainAxisAlignment.end,
        columns: [0.2, 0.3, 0.3, 0.2],
        rows: [0.2, 0.3, 0.3, 0.2],
        children: [
          GridItem(
            start: const Size(0, 0),
            end: const Size(4, 1),
            child: Container(color: Colors.red),
          ),
          GridItem(
            start: const Size(1, 0),
            end: const Size(3, 4),
            order: 1,
            child: Container(color: Colors.blue.withOpacity(0.5)),
          ),
          GridItem(
            start: const Size(2, 3),
            end: const Size(4, 4),
            child: Container(color: Colors.green),
          ),
        ],
      ),
    ];

    for (int i = 0; i < testCases.length; i++) {
      testWidgets(testCases[i].title, (WidgetTester tester) async {
        await tester.pumpWidget(
          MaterialApp(
            debugShowCheckedModeBanner: false,
            theme: ThemeData(
              brightness: Brightness.light,
              useMaterial3: true,
            ),
            home: RepaintBoundary(
              child: Container(
                decoration: const BoxDecoration(color: Colors.yellow),
                child: GridContainer(
                  alignment: testCases[i].alignment,
                  columns: testCases[i].columns.cast(),
                  rows: testCases[i].rows.cast(),
                  children: testCases[i].children.cast(),
                ),
              ),
            ),
          ),
        );
        // flutter test --update-goldens
        await expectLater(
          find.byType(RepaintBoundary).first,
          matchesGoldenFile('./grid_container_test.dart.$i.png'),
        );
      });
    }
  });
}


example/lib/basic_page.dart
// Copyright 2023 The terCAD team. All rights reserved.
// Use of this source code is governed by a MIT license
// that can be found in the LICENSE file.

import 'dart:math';

import 'package:flutter/material.dart';
import 'components_builder.dart';

typedef ComponentData = Map<String, dynamic>;

abstract class InterfaceComponent {
  static const key = '_component';
  static const order = '_order';
  static const endX = '_end_x';
  static const endY = '_end_y';
  static const startX = '_start_x';
  static const startY = '_start_y';
  static const start = '_start';
  static const end = '_end';
  static const shift = '_shift';
  static const color = 'color';
}

class BasicPage extends StatefulWidget {
  const BasicPage({super.key});

  @override
  BasicPageState createState() => BasicPageState();
}

class BasicPageState extends State<BasicPage> {
  List<ComponentData> data = [];
  late String key;

  Future<void> _save() async {
    // ... store data
    setState(() {});
  }

  Future<void> save() async {
    await _save();
  }

  Future<void> drop() async {
    data.clear();
    // ... update store
  }

  Future<void> add(String key) async {
    List<Color> colors = Colors.primaries;
    Random random = Random();
    data.add({
      InterfaceComponent.key: key,
      InterfaceComponent.startX: 0,
      InterfaceComponent.startY: 0,
      InterfaceComponent.endX: 1,
      InterfaceComponent.endY: 1,
      InterfaceComponent.color: colors[random.nextInt(colors.length)],
    });
    await _save();
  }

  Future<void> adjust(int index, ComponentData change) async {
    data[index] = change;
    await _save();
  }

  Future<void> delete(int index) async {
    data.removeAt(index);
    await _save();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        toolbarHeight: 45,
        backgroundColor: Colors.blueGrey,
        title: const Text('Layout Manager'),
        centerTitle: true,
        leadingWidth: 120,
        leading: Row(
          children: [
            IconButton(
              hoverColor: Colors.transparent,
              icon: const Icon(
                Icons.save,
                color: Colors.white70,
              ),
              tooltip: 'Save',
              onPressed: save,
            ),
            IconButton(
              hoverColor: Colors.transparent,
              icon: const Icon(
                Icons.cancel,
                color: Colors.white70,
              ),
              tooltip: 'Delete',
              onPressed: drop,
            ),
          ],
        ),
        actions: [
          IconButton(
            hoverColor: Colors.transparent,
            icon: const Icon(
              Icons.add_circle_rounded,
              color: Colors.white70,
            ),
            tooltip: 'Add',
            onPressed: () => add(UniqueKey().toString()),
          ),
        ],
      ),
      body: Padding(
        padding: const EdgeInsets.all(12),
        child: ComponentsBuilder(data,
            editMode: true, adjust: adjust, delete: delete),
      ),
    );
  }
}


lib/src/grid_container.dart
// Copyright 2023 The terCAD team. All rights reserved.
// Use of this source code is governed by a MIT license
// that can be found in the LICENSE file.

import 'package:flutter/widgets.dart';
import 'package:flutter_grid_layout/flutter_grid_layout.dart';
import 'package:flutter_grid_layout/src/list_ext.dart';

class GridContainer extends StatelessWidget {
  /// Columns size declaration
  // 0.5 -> 50%, 10 -> 10 DIP, null -> auto
  final List<double?> columns;

  /// Rows size declaration
  // 0.5 -> 50%, 10 -> 10 DIP, null -> auto
  final List<double?> rows;

  /// List of elements for plotting
  final List<GridItem> children;

  /// Type of alignment
  // Right-to-Left: MainAxisAlignment.end
  // Left-to-Right: MainAxisAlignment.start
  final MainAxisAlignment alignment;

  const GridContainer({
    super.key,
    required this.children,
    required this.columns,
    required this.rows,
    this.alignment = MainAxisAlignment.start,
  });

  // Calculate DIP for relative values
  List<double> _calc(double size, List<double?> scope) {
    int restCount = scope.where((e) => e == null).length;
    double takenSize = 0;
    for (int i = 0; i < scope.length; i++) {
      if (scope[i] == null) {
        continue;
      }
      double value = scope[i]!;
      if (value > 0 && value < 1) {
        scope[i] = value * size;
        takenSize += scope[i]!;
      } else {
        takenSize += value;
      }
    }
    if (takenSize > size) {
      double cut = (size - takenSize) / (scope.length - restCount);
      scope = scope.map((value) => value != null ? value + cut : null).toList();
      takenSize = size;
    }
    if (restCount > 0) {
      double rest = (size - takenSize) / restCount;
      scope = scope.map((value) => value ?? rest).toList();
    }
    scope.insert(0, 0.0);
    return scope.cast();
  }

  // Restate widths from zero-point
  List<double> _scale(List<double> scope) {
    return scope
        .asMap()
        .entries
        .map((entry) =>
            scope.sublist(0, entry.key + 1).fold(0.0, (v, e) => v + e))
        .toList();
  }

  // Calculate widths from relative values
  List<double> _calcWidth(double maxWidth) {
    if (alignment == MainAxisAlignment.end) {
      return _calc(maxWidth, rows.reversed.toList());
    }
    return _calc(maxWidth, rows);
  }

  // Calculate heights from relative values
  List<double> _calcHeight(double maxHeight) => _calc(maxHeight, columns);

  GridItem _get(index) {
    GridItem item = children[index];
    if (alignment == MainAxisAlignment.end) {
      item = GridItem(
        order: item.zIndex,
        start: Size(rows.length - item.end.width, item.start.height),
        end: Size(rows.length - item.start.width, item.end.height),
        child: item.child,
      );
    }
    return item;
  }

  @override
  Widget build(BuildContext context) {
    if (columns.isEmpty || rows.isEmpty || children.isEmpty) {
      return const SizedBox();
    }
    children.sort((a, b) => a.zIndex.compareTo(b.zIndex));
    return LayoutBuilder(builder: (context, constraints) {
      final width = _scale(_calcWidth(constraints.maxWidth));
      final height = _scale(_calcHeight(constraints.maxHeight));
      return Stack(
        children: List<Widget>.generate(children.length, (index) {
          final item = _get(index);
          final itemWidth =
              width.by(item.end.width) - width.by(item.start.width);
          final itemHeight =
              height.by(item.end.height) - height.by(item.start.height);
          if (itemWidth <= 0 || itemHeight <= 0) {
            return const SizedBox();
          }
          return Container(
            margin: EdgeInsets.only(
              left: width.by(item.start.width),
              top: height.by(item.start.height),
            ),
            constraints: BoxConstraints(
              maxWidth: itemWidth,
              minWidth: itemWidth,
              maxHeight: itemHeight,
              minHeight: itemHeight,
            ),
            width: itemWidth,
            height: itemHeight,
            child: item.child,
          );
        }),
      );
    });
  }
}


example/lib/components_builder.dart
// Copyright 2023 The terCAD team. All rights reserved.
// Use of this source code is governed by a MIT license
// that can be found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:flutter_grid_layout/flutter_grid_layout.dart';
import 'basic_page.dart';

class ComponentsBuilder extends StatelessWidget {
  final List<ComponentData> data;
  final bool editMode;
  final Function? adjust;
  final Function? delete;
  final _shift = InterfaceComponent.shift;
  final _order = InterfaceComponent.order;
  final _start = InterfaceComponent.start;
  final _end = InterfaceComponent.end;

  const ComponentsBuilder(
    this.data, {
    super.key,
    this.editMode = false,
    this.adjust,
    this.delete,
  });

  void resize(ComponentData change, Size start) {
    final scope = data[change[_order]];
    scope[_order] = change[_order];
    if (change[_shift] != null) {
      scope[InterfaceComponent.endX] +=
          start.width - scope[InterfaceComponent.startX];
      scope[InterfaceComponent.endY] +=
          start.height - scope[InterfaceComponent.startY];
      scope[InterfaceComponent.startX] = start.width;
      scope[InterfaceComponent.startY] = start.height;
    } else if (change[_start] != null) {
      scope[InterfaceComponent.startX] = start.width;
      scope[InterfaceComponent.startY] = start.height;
    } else if (change[_end] != null) {
      scope[InterfaceComponent.endX] = start.width + 1.0;
      scope[InterfaceComponent.endY] = start.height + 1.0;
    }
    adjust!(scope[_order], scope);
  }

  @override
  Widget build(BuildContext context) {
    const rowsCount = 6;
    const columnsCount = 6;
    return GridContainer(
      rows: List.filled(rowsCount, null),
      columns: List.filled(columnsCount, null),
      children: editMode
          ? [
              ...List.generate(rowsCount * columnsCount, (i) {
                final start = Size(
                  i % rowsCount.toDouble(),
                  (i ~/ rowsCount).toDouble(),
                );
                return GridItem(
                  start: start,
                  end: Size(start.width + 1, start.height + 1),
                  order: 0,
                  child: DragTarget<ComponentData>(
                    onWillAccept: (_) => true,
                    onAccept: (change) => resize(change, start),
                    builder: (context, candidateData, _) => Container(
                      decoration: BoxDecoration(
                        border: Border.all(color: Colors.grey),
                        color: candidateData.isEmpty
                            ? null
                            : Colors.green.shade200,
                      ),
                    ),
                  ),
                );
              }),
              ...List.generate(
                data.length,
                (i) => GridItem(
                  start: Size(data[i][InterfaceComponent.startX] + .0,
                      data[i][InterfaceComponent.startY] + .0),
                  end: Size(data[i][InterfaceComponent.endX] + .0,
                      data[i][InterfaceComponent.endY] + .0),
                  order: i + 1,
                  child: Draggable<ComponentData>(
                    data: {...data[i], _order: i, _shift: true},
                    feedback: Container(
                      color: Colors.amberAccent,
                      width: 32,
                      height: 24,
                    ),
                    child: Container(color: data[i][InterfaceComponent.color]),
                  ),
                ),
              ),
            ]
          : List.generate(
              data.length,
              (i) => GridItem(
                    start: Size(data[i][InterfaceComponent.startX] + .0,
                        data[i][InterfaceComponent.startY] + .0),
                    end: Size(data[i][InterfaceComponent.endX] + .0,
                        data[i][InterfaceComponent.endY] + .0),
                    child: const SizedBox(),
                  )),
    );
  }
}


example/lib/main.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:flutter_grid_layout/flutter_grid_layout.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(const MyApp());
}

// Serialized component state
typedef ComponentData = Map<String, dynamic>;

// Properties for ComponentData
abstract class InterfaceComponent {
  static const key = '_component';
  static const order = '_order';
  static const endX = '_end_x';
  static const endY = '_end_y';
  static const startX = '_start_x';
  static const startY = '_start_y';
  static const start = '_start';
  static const end = '_end';
  static const shift = '_shift';
  static const color = 'color';
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  MyAppState createState() => MyAppState();
}

class MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        brightness: Brightness.light,
        useMaterial3: true,
      ),
      darkTheme: ThemeData(
        brightness: Brightness.dark,
        useMaterial3: true,
      ),
      themeMode: ThemeMode.system,
      home: const BasicPage(),
    );
  }
}

class BasicPage extends StatefulWidget {
  const BasicPage({super.key});

  @override
  BasicPageState createState() => BasicPageState();
}

class BasicPageState extends State<BasicPage> {
  List<ComponentData> data = [];
  late String key;

  Future<void> _save() async {
    // ... store data
    setState(() {});
  }

  Future<void> save() async {
    // ... other actions with `data`
    await _save();
  }

  Future<void> drop() async {
    data.clear();
    // ... update store
  }

  Future<void> add(String key) async {
    List<Color> colors = Colors.primaries;
    Random random = Random();
    data.add({
      InterfaceComponent.key: key,
      InterfaceComponent.startX: 0,
      InterfaceComponent.startY: 0,
      InterfaceComponent.endX: 1 + random.nextInt(3),
      InterfaceComponent.endY: 1 + random.nextInt(2),
      InterfaceComponent.color: colors[random.nextInt(colors.length)],
    });
    await _save();
  }

  Future<void> adjust(int index, ComponentData change) async {
    data[index] = change;
    await _save();
  }

  Future<void> delete(int index) async {
    data.removeAt(index);
    await _save();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        toolbarHeight: 45,
        backgroundColor: Colors.blueGrey,
        title: const Text('Layout Manager'),
        centerTitle: true,
        leadingWidth: 120,
        leading: Row(
          children: [
            IconButton(
              hoverColor: Colors.transparent,
              icon: const Icon(
                Icons.save,
                color: Colors.white70,
              ),
              tooltip: 'Save',
              onPressed: save,
            ),
            IconButton(
              hoverColor: Colors.transparent,
              icon: const Icon(
                Icons.cancel,
                color: Colors.white70,
              ),
              tooltip: 'Delete',
              onPressed: drop,
            ),
          ],
        ),
        actions: [
          IconButton(
            hoverColor: Colors.transparent,
            icon: const Icon(
              Icons.add_circle_rounded,
              color: Colors.white70,
            ),
            tooltip: 'Add',
            onPressed: () => add(UniqueKey().toString()),
          ),
        ],
      ),
      body: Padding(
        padding: const EdgeInsets.all(12),
        child: ComponentsBuilder(data,
            editMode: true, adjust: adjust, delete: delete),
      ),
    );
  }
}

class ComponentsBuilder extends StatelessWidget {
  final List<ComponentData> data;
  final bool editMode;
  final Function? adjust;
  final Function? delete;
  final _shift = InterfaceComponent.shift;
  final _order = InterfaceComponent.order;
  final _start = InterfaceComponent.start;
  final _end = InterfaceComponent.end;

  const ComponentsBuilder(
    this.data, {
    super.key,
    this.editMode = false,
    this.adjust,
    this.delete,
  });

  void resize(ComponentData change, Size start) {
    final scope = data[change[_order]];
    scope[_order] = change[_order];
    if (change[_shift] != null) {
      scope[InterfaceComponent.endX] +=
          start.width - scope[InterfaceComponent.startX];
      scope[InterfaceComponent.endY] +=
          start.height - scope[InterfaceComponent.startY];
      scope[InterfaceComponent.startX] = start.width;
      scope[InterfaceComponent.startY] = start.height;
    } else if (change[_start] != null) {
      scope[InterfaceComponent.startX] = start.width;
      scope[InterfaceComponent.startY] = start.height;
    } else if (change[_end] != null) {
      scope[InterfaceComponent.endX] = start.width + 1.0;
      scope[InterfaceComponent.endY] = start.height + 1.0;
    }
    adjust!(scope[_order], scope);
  }

  @override
  Widget build(BuildContext context) {
    const rowsCount = 6;
    const columnsCount = 6;
    return GridContainer(
      rows: List.filled(rowsCount, null),
      columns: List.filled(columnsCount, null),
      children: editMode
          ? [
              ...List.generate(rowsCount * columnsCount, (i) {
                final start = Size(
                  i % rowsCount.toDouble(),
                  (i ~/ rowsCount).toDouble(),
                );
                return GridItem(
                  start: start,
                  end: Size(start.width + 1, start.height + 1),
                  order: 0,
                  child: DragTarget<ComponentData>(
                    onWillAccept: (_) => true,
                    onAccept: (change) => resize(change, start),
                    builder: (context, candidateData, _) => Container(
                      decoration: BoxDecoration(
                        border: Border.all(color: Colors.grey),
                        color: candidateData.isEmpty
                            ? null
                            : Colors.green.shade200,
                      ),
                    ),
                  ),
                );
              }),
              ...List.generate(
                data.length,
                (i) => GridItem(
                  start: Size(data[i][InterfaceComponent.startX] + .0,
                      data[i][InterfaceComponent.startY] + .0),
                  end: Size(data[i][InterfaceComponent.endX] + .0,
                      data[i][InterfaceComponent.endY] + .0),
                  order: i + 1,
                  child: Draggable<ComponentData>(
                    data: {...data[i], _order: i, _shift: true},
                    feedback: Container(
                      color: Colors.amberAccent,
                      width: 32,
                      height: 24,
                    ),
                    child: Container(color: data[i][InterfaceComponent.color]),
                  ),
                ),
              ),
            ]
          : List.generate(
              data.length,
              (i) => GridItem(
                    start: Size(data[i][InterfaceComponent.startX] + .0,
                        data[i][InterfaceComponent.startY] + .0),
                    end: Size(data[i][InterfaceComponent.endX] + .0,
                        data[i][InterfaceComponent.endY] + .0),
                    child: const SizedBox(), // STUB
                  )),
    );
  }
}
